RI: Get user tags only from db if size > 0

ashiagr opened the pull request:
This PR 

1. updates`ReaderTagRepository.getUserTags()` to return tags only from the db if count > 0 - d1e5582

    If there are already followed tags in the db, we can skip fetching user tags from the server (and avoid unnecessary full screen error message due to n/w error) as the new discover feed can be populated based on these tags.

2. deletes recommended tags added as followed tags for logged out users inside `GetFollowedTagsUseCase.get()` - fd6c575

    This is done so that they don't cause issues in calculating followed tags count during display of "Choose Interest" screen. 
    More context [here](https://github.com/wordpress-mobile/WordPress-Android/pull/12530#discussion_r462794836).

3.  fixes a loading issue in the discover feed - ef28a4e

#### To Test

Enable RI FF 

1. Launch app and login with account with followed tags
2. Enable airplane mode
3. Go to Reader -> Discover tab 
4. Notice that discover feed is shown and not blocked by any full screen error message

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Fetch user tags directly from db

ashiagr committed:
Check if state is correct before accessing content

malinajirka committed:
Work-in progress commit - SHOULD NOT BE IN DEVELOP

ashiagr committed:
Revert "Fetch user tags directly from db"

This reverts commit a381df53

ashiagr committed:
Merge branch 'malinjir-user-tags-fix' into issue/get-user-tags-from-db

ashiagr committed:
Extract tags deletion from ReaderUpdateLogic to ReaderTagTable

ashiagr committed:
Check if followed tags not empty in db then return

ashiagr committed:
Delete recommended tags for logged out users

Earlier these were added as followed tags for a logged out users. Once we've enabled reader improvements feature, which allows tags to be followed locally for logged out user, these tags will not be needed anymore.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/datasets/ReaderTagTable.java:
```diff
@@ -363,4 +372,31 @@ private static int minutesSinceLastUpdate(ReaderTag tag) {
+        db.beginTransaction();
+        try {
+            for (ReaderTag tag : tagList) {
+                ReaderPostTable.deletePostsWithTag(tag);
```
Do we really need to delete the posts here? It feels a bit weird to access one table from another. I think "dangling" posts will get deleted automatically [here](https://github.com/wordpress-mobile/WordPress-Android/blob/f7760b32bd2cc702618858e1b941c2d92089505d/WordPress/src/main/java/org/wordpress/android/datasets/ReaderPostTable.java#L215). Wdyt?

ashiagr replied:
It was an existing method that I moved from `ReaderUpdateLogic` to the `ReaderTagTable`.  I suppose posts deletion can be handled at the line you linked.

malinajirka replied:
If we remove the tags in the migration script I believe we won't need this method, right?

ashiagr replied:
Removed this method and used `deleteTags() `directly when local tags are synced with the server in in 5c6f8e3.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/datasets/ReaderTagTable.java:
```diff
@@ -363,4 +372,31 @@ private static int minutesSinceLastUpdate(ReaderTag tag) {
         return DateTimeUtils.minutesBetween(dtUpdated, dtNow);
     }
+
+    public static void deleteRecommendedTagsAddedAsFollowedTags() {
```
I'm wondering if we could delete the out-dated data in a migration script (when we get rid of the FF tomorrow). We could drop the whole tag table when it comes to that - it's just a cache anyway. This code will be here forever and every user will use it once at most. Wdyt? (I realize it technically wasn't introduced in this PR, but imo it's still worth considering)

ashiagr replied:
Yep, we can include it in migration script. Wasn't sure how to migrate with RI FF enabled.

malinajirka replied:
yeah, I agree that'd be quite tricky with a FF. Feel free to create the migration script ignoring the FF - we'll delete the tags for everyone no matter the FF. I'll create a PR for removing the FF tomorrow anyway. Thanks 🙇

ashiagr replied:
Deleted tags for everyone in f493d12. 
Removed now unused `deleteRecommendedTagsAddedAsFollowedTags()` in 8813d3e.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/tags/GetFollowedTagsUseCase.kt:
```diff
@@ -3,16 +3,23 @@ package org.wordpress.android.ui.reader.repository.usecases.tags
+    private val appPrefsWrapper: AppPrefsWrapper
 ) {
     suspend fun get(): ReaderTagList = withContext(ioDispatcher) {
+        if (!accountStore.hasAccessToken() && appPrefsWrapper.isReaderImprovementsPhase2Enabled()) {
```
It feels a bit out of place here - we can discuss it further based on the above comments. Ignore this comment for now.

malinajirka reviewed (commented):
Thanks @ashiagr! I've done a first review pass. Let me know what you think.

ashiagr committed:
Remove deleteTagsAndPostsWithTags

"dangling" posts will get deleted automatically inside ReaderPostTable.purge

ashiagr committed:
Remove deleteRecommendedTagsAddedAsFollowedTags

This was added to remove recommended tags added as followed tags while RI FF was enabled. With RI FF being removed, we can now write a migration script to delete tags and remove this method.

ashiagr committed:
Delete followed tags in migration

ashiagr committed:
Merge branch 'develop' into issue/get-user-tags-from-db

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderFragment.kt

malinajirka reviewed (approved):
I've tested the migration with both a self-hosted and .com sites. Everything works as expected. Thanks for all the work Ashita;)!
